client/asset/eth/eth.go

```diff
@@ -1605,29 +2661,204 @@ func (eth *ExchangeWallet) redeem(ctx context.Context, redemptions []*asset.Rede
 // refund refunds a swap contract using the account controlled by the wallet.
 // Any on-chain failure, such as the locktime not being past, will not cause
 // this to error.
-func (eth *ExchangeWallet) refund(ctx context.Context, secretHash [32]byte, maxFeeRate uint64, contractVer uint32) (tx *types.Transaction, err error) {
-	eth.nonceSendMtx.Lock()
-	defer eth.nonceSendMtx.Unlock()
-	gas := dexeth.RefundGas(contractVer)
-	txOpts, _ := eth.node.txOpts(ctx, 0, gas, dexeth.GweiToWei(maxFeeRate))
-	return tx, eth.withContractor(contractVer, func(c contractor) error {
+func (w *AssetWallet) refund(secretHash [32]byte, maxFeeRate uint64, contractVer uint32) (tx *types.Transaction, err error) {
+	gas := w.gases(contractVer)
+	if gas == nil {
+		return nil, fmt.Errorf("no gas table for asset %d, version %d", w.assetID, contractVer)
+	}
+	w.nonceSendMtx.Lock()
+	defer w.nonceSendMtx.Unlock()
+	txOpts, err := w.node.txOpts(w.ctx, 0, gas.Refund, dexeth.GweiToWei(maxFeeRate))
+	if err != nil {
+		return nil, fmt.Errorf("addSignerToOpts error: %w", err)
+	}
+
+	return tx, w.withContractor(contractVer, func(c contractor) error {
 		tx, err = c.refund(txOpts, secretHash)
 		return err
 	})
 }
 
 // isRedeemable checks if the swap identified by secretHash is redeemable using secret.
-func (eth *ExchangeWallet) isRedeemable(secretHash [32]byte, secret [32]byte, contractVer uint32) (redeemable bool, err error) {
-	return redeemable, eth.withContractor(contractVer, func(c contractor) error {
+func (w *AssetWallet) isRedeemable(secretHash [32]byte, secret [32]byte, contractVer uint32) (redeemable bool, err error) {
+	return redeemable, w.withContractor(contractVer, func(c contractor) error {
 		redeemable, err = c.isRedeemable(secretHash, secret)
 		return err
 	})
 }
 
-// isRefundable checks if the swap identified by secretHash is refundable.
-func (eth *ExchangeWallet) isRefundable(secretHash [32]byte, contractVer uint32) (refundable bool, err error) {
-	return refundable, eth.withContractor(contractVer, func(c contractor) error {
+func (w *AssetWallet) isRefundable(secretHash [32]byte, contractVer uint32) (refundable bool, err error) {
+	return refundable, w.withContractor(contractVer, func(c contractor) error {
 		refundable, err = c.isRefundable(secretHash)
 		return err
 	})
 }
+
+// getGasEstimate is used to get a gas table for an asset's contract(s). The
+// provided gases, g, should be generous estimates of what the gas might be.
+// Errors are thrown if the provided estimates are too small by more than a
+// factor of 2. The account should already have a trading balance of at least
+// maxSwaps gwei (token or eth), and sufficient eth balance to cover the
+// requisite tx fees.
+func GetGasEstimates(ctx context.Context, cl *nodeClient, c contractor, maxSwaps int, g *dexeth.Gases, waitForMined func()) error {
```

Comment:
Comment is unexported, but this is exported. Will this be part of another tool? Does it belong in testing? Uses waitForMined() which is part of the harness tests. Not sure if you can use functions from test files in the build.

Reply:
`waitForMined` is passed in as an argument for now. This will be used to create the gas tables once a contract is launched on a network. I haven't written a utility for it yet, but I'm planning on it.